scan/base.go

```diff
@@ -40,6 +42,267 @@ type base struct {
+	if themeVuln, err = detectWpTheme(c); err != nil {
+		return
+	}
+	for _, i := range themeVuln {
+		rs = append(rs, i)
+	}
+
+	var pluginVuln []models.VulnInfo
+	if pluginVuln, err = detectWpPlugin(c); err != nil {
+		return
+	}
+	for _, i := range pluginVuln {
+		rs = append(rs, i)
+	}
+	return
+}
+
+func detectWpCore(c config.ServerInfo) (rs []models.VulnInfo, err error) {
+	cmd := fmt.Sprintf("wp core version --path=%s", c.WpPath)
+
+	var coreVersion string
+	if r := exec(c, cmd, noSudo); r.isSuccess() {
+		tmp := strings.Split(r.Stdout, ".")
+		coreVersion = strings.Join(tmp, "")
+		coreVersion = strings.TrimRight(coreVersion, "\r\n")
+		if len(coreVersion) == 0 {
+			return
+		}
+	}
+	cmd = fmt.Sprintf("curl -H 'Authorization: Token token=%s' https://wpvulndb.com/api/v3/wordpresses/%s", c.WpToken, coreVersion)
+	if r := exec(c, cmd, noSudo); r.isSuccess() {
+		data := map[string]WpCveInfos{}
+		if err = json.Unmarshal([]byte(r.Stdout), &data); err != nil {
+			return
+		}
+		for _, i := range data {
+			if len(i.Vulnerabilities) == 0 {
+				continue
+			}
+			for _, e := range i.Vulnerabilities {
+				var cveIDs []string
+				for _, k := range e.References.Cve {
+					cveIDs = append(cveIDs, "CVE-"+k)
+				}
+
+				for _, cveID := range cveIDs {
+					rs = append(rs, models.VulnInfo{
+						CveID: cveID,
+					})
+				}
+			}
+		}
+	}
+	return
+}
+
+//WpStatus is for wp command
+type WpStatus struct {
+	Name    string `json:"-"`
+	Status  string `json:"-"`
+	Update  string `json:"-"`
+	Version string `json:"-"`
+}
+
+func detectWpTheme(c config.ServerInfo) (rs []models.VulnInfo, err error) {
+	cmd := fmt.Sprintf("wp theme list --path=%s", c.WpPath)
+
+	var themes []WpStatus
+	if r := exec(c, cmd, noSudo); r.isSuccess() {
+		themes = parseStatus(r.Stdout)
+	}
+
+	for _, theme := range themes {
+		cmd := fmt.Sprintf("curl -H 'Authorization: Token token=%s' https://wpvulndb.com/api/v3/themes/%s", c.WpToken, theme.Name)
+		if r := exec(c, cmd, noSudo); r.isSuccess() {
+			data := map[string]WpCveInfos{}
+			if err = json.Unmarshal([]byte(r.Stdout), &data); err != nil {
+				var jsonError WpError
+				if err = json.Unmarshal([]byte(r.Stdout), &jsonError); err != nil {
+					return
+				}
+				continue
+			}
+			for _, i := range data {
+				if len(i.Vulnerabilities) == 0 {
+					continue
+				}
+				if len(i.Error) != 0 {
+					continue
+				}
+				for _, e := range i.Vulnerabilities {
+					v1, _ := version.NewVersion(theme.Version)
+					v2, _ := version.NewVersion(e.FixedIn)
+					if v1.LessThan(v2) {
+						if len(e.References.Cve) == 0 {
+							continue
+						}
+						var cveIDs []string
+						for _, k := range e.References.Cve {
+							cveIDs = append(cveIDs, "CVE-"+k)
+						}
+
+						for _, cveID := range cveIDs {
+							rs = append(rs, models.VulnInfo{
+								CveID: cveID,
+							})
+						}
+					}
+				}
+			}
+
+		}
+	}
+	return
+}
+
+type Errors struct {
+	Error string `json:"error"`
+}
+
+type WpError struct {
```

Comment:
__[Golint]__ exported type WpError should have comment or be unexported

<sub>posted by Sider</sub>

Reply:
implement